### PR TITLE
add `DDEV_DATABASE_FAMILY` variable, fixes #4954

### LINKED ISSUE #4954: Expose database type as variable for use in scripts
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

DDEV supports MySQL/Maria and Postgres databases.

However, DDEV does not directly expose the database type (mysql, postgres) for use in scripts and addons.
This prevents developers from easily generating database URLs programatically: EG. `mysql://db:db@db/db`

I noticed a recent [update ](https://github.com/ddev/ddev-drupal-contrib/blob/main/config.contrib.yaml) to [ddev/ddev-drupal-contrib](https://github.com/ddev/ddev-drupal-contrib) sets a variable for testing:

```yaml
SIMPLETEST_DB=mysql://db:db@db/db
```
 
My company has standardized on postgres and I would like match the testing env to production env.

### Describe your solution

I would like to be able to re-write the code to the following:

```diff
-  - SIMPLETEST_DB=mysql://db:db@db/db
+  - SIMPLETEST_DB=${DDEV_DATABASE_TYPE}://db:db@db/db
```

`DDEV_DATABASE_TYPE` should be either:
- `mysql` for MySql and Maria dbs
- `postgres` for [Postgres dbs](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING)

### Describe alternatives

I am aware that I can remove the `ddev_generated` text and make the changes myself to the file. 
But this prevents me getting the latest changes and must be done manually for each project.

### Additional context

If this feature request is accepted, it would be nice if addons could overrite the setting somehow.
There is also a [Mongo database addon](https://github.com/ddev/ddev-mongo) so it should be able to set `mongo` as a valid type.

### PR DESCRIPTION
**Pull request by tyler36**
## The Issue

* #4954 

I want to generate database resource URLs for use in code and DDEV addons.

Currently, DDEV exposes `DDEV_DATABASE` which is  `app.Database.Type + ":" + app.Database.Version,`.

This is difficult to use because it must be parsed, and converted (MariaDB uses `mysql` protocol).

## How This PR Solves The Issue

This PR adds a new environmental variable: `DDEV_DATABASE_FAMILY`

- Defaults to `mysql`
- Set to `postgres` if database is `postgres`

## Manual Testing Instructions

1. Start a DDEV project
1. Type `ddev exec set | grep DDEV_DATABASE` to check currently set database variables

## Related Issue Link(s)

Fixes #4954

## Release/Deployment Notes

<a href="https://gitpod.io/#https://github.com/ddev/ddev/pull/4991"></a>

### CODE REVIEW

**Comment by rfay** [comment]
Look at you go!

**Comment by tyler36** [comment]
Sorry for the noise. Finally added a "GO" addon to VScode for proper syntax highlighting etc.

**Review by rfay — CHANGES_REQUESTED** [review]
Thanks! 

* Still needs to be added to docs (custom_commands.md) If you find other things missing there, OK to add.
* Still needs to be added to test, TestEnvironmentVariables (you may find that some other things like DDEV_DATABASE are missing from test, OK to add.

File: pkg/ddevapp/ddevapp.go [hunk 1]
```diff
@@ -1958,6 +1958,14 @@ func (app *DdevApp) DockerEnv() {
 		dbPortStr = app.HostDBPort
 	}
 
+	// DDEV_DATABASE_PROTOCOL can be use for connection URLs in the format of protocol//[hosts][/database][?properties]
```
Comment (gilbertsoft) [inline pkg/ddevapp/ddevapp.go:1961]:
```suggestion
	// DDEV_DATABASE_PROTOCOL can be use for connection URLs
```
I'd remove this part as it's anyway not complete, colon and auth part are missing. But should be clear with the following example.

**Review by gilbertsoft — CHANGES_REQUESTED** [review]
Thanks! Mainly only a minor naming issue.

And you have to add the new variable to the docker-composer template `pkg/ddevapp/app_compose_template.yaml`.

File: pkg/ddevapp/ddevapp.go (hunk 1 above)
Comment (tyler36) [reply pkg/ddevapp/ddevapp.go]: Fixed in 7742323ad8183b0785fdfdae5e9530325365a897

File: pkg/ddevapp/ddevapp.go [hunk 2]
```diff
@@ -1961,6 +1969,7 @@ func (app *DdevApp) DockerEnv() {
 		"DDEV_WEBIMAGE":                  app.WebImage,
 		"DDEV_APPROOT":                   app.AppRoot,
 		"DDEV_COMPOSER_ROOT":             app.GetComposerRoot(true, false),
+		"DDEV_DATABASE_PROTOCOL":         dbProtocol,
```
[the remark this answers is not in the document]
Comment (tyler36) [reply pkg/ddevapp/ddevapp.go]: Fixed in 064945ad1d8201912a50072fe29b49f85a27c1b5

**Comment by tyler36** [comment]
> And you have to add the new variable to the docker-composer template pkg/ddevapp/app_compose_template.yaml.

OK. That was the piece I was missing.

**Comment by tyler36** [comment]
> Error: File is not `gofmt`-ed with `-s` (gofmt)

Not sure why the golint test is failing.

**Comment by gilbertsoft** [comment]
> > Error: File is not `gofmt`-ed with `-s` (gofmt)
> 
> Not sure why the golint test is failing.

You just miss two spaces at the mentioned line.

File: pkg/ddevapp/ddevapp.go [hunk 3]
```diff
@@ -1971,6 +1979,7 @@ func (app *DdevApp) DockerEnv() {
 		"DDEV_WEBIMAGE":                  app.WebImage,
 		"DDEV_APPROOT":                   app.AppRoot,
 		"DDEV_COMPOSER_ROOT":             app.GetComposerRoot(true, false),
+		"DDEV_DATABASE_FAMILY":         dbFamily,
```
Comment (gilbertsoft) [inline pkg/ddevapp/ddevapp.go:1982]:
```suggestion
		"DDEV_DATABASE_FAMILY":           dbFamily,
```

**Review by gilbertsoft — CHANGES_REQUESTED** [review]

File: pkg/ddevapp/app_compose_template.yaml [hunk 4]
```diff
@@ -45,6 +45,7 @@ services:
     environment:
       - COLUMNS
       - DDEV_DATABASE
+      - DDEV_DATABASE_FAMILY
```
Comment (gilbertsoft) [inline pkg/ddevapp/app_compose_template.yaml:48]: Also add it bellow to the web container (~line 172)

**Review by gilbertsoft — CHANGES_REQUESTED** [review]

File: pkg/ddevapp/app_compose_template.yaml (hunk 4 above)
Comment (tyler36) [reply pkg/ddevapp/app_compose_template.yaml:48]: Fixed in #0203dcc6bd8d1382d7507d47707e4a88aa91aab2

**Review by gilbertsoft — APPROVED** [review]
Looks good. Restarted the failing test which should become green too...

**Review by rfay — APPROVED** [review]
Looks great to me, thanks! Manually tested fine.

**Comment by rfay** [comment]
Congratulations and thanks!

**Pull request merged**